File: Lista 2 - Ad-Hoc e ordenação e Estrutura de Dados/10  - Train Swapping.py
def swap(train, a, b):
    temp = train[a]
    train[a] = train[b]
    train[b] = temp

def sort_train(train):
    swaps = 0

    while True:
        something_changed = False

        for i in range(len(train) - 1):
            if train[i] > train[i + 1]:
                something_changed = True
                swaps += 1
                swap(train, i, i+1)

        if not something_changed:
            break
    
    return swaps


def test():
    n = int(input())

    for _ in range(n):
        input()
        trains = [int(x) for x in input().split()]
        
        #print(trains)
        swaps = sort_train(trains)
        #print(trains)
        
        print("Optimal train swapping takes {} swaps.".format(swaps))

#test()

test()

File: Lista 2 - Ad-Hoc e ordenação e Estrutura de Dados/10  - Train Swapping.cpp
#include <iostream>

using namespace std;

void swap(int input_list[], int index1, int index2) {
    int temp = input_list[index1];
    input_list[index1] = input_list[index2];
    input_list[index2] = temp;
}

int sort_train(int train[], int l) {
    int swaps = 0;
    bool something_changed;

    do {

        something_changed = false;

        for(int j = 0; j < (l - 1); j++){
            if(train[j] > train[j+1]){
                something_changed = true;
                swap(train, j, j+1);
                swaps++;
            }
        }

    } while(something_changed);

    return  swaps;
}

int main() {
    int n, l;
    int train[50];
    int swaps;

    scanf(" %d", &n);

    for(int i = 0; i < n; i++){
        scanf(" %d", &l);
        
        for(int j = 0; j < l; j++){
            scanf(" %d", &train[j]);
        }
        
        swaps = sort_train(train, l);
        printf("Optimal train swapping takes %d swaps.\n", swaps);
    }


    return 0;
}